src/ray/object_manager/plasma/object_lifecycle_manager.cc

```diff
@@ -0,0 +1,281 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "ray/object_manager/plasma/object_lifecycle_manager.h"
+#include "absl/time/clock.h"
+#include "ray/common/ray_config.h"
+
+namespace plasma {
+using namespace flatbuf;
+
+ObjectLifecycleManager::ObjectLifecycleManager(
+    IAllocator &allocator, ray::DeleteObjectCallback delete_object_callback)
+    : object_store_(std::make_unique<ObjectStore>(allocator)),
+      eviction_policy_(std::make_unique<EvictionPolicy>(*object_store_, allocator)),
+      delete_object_callback_(delete_object_callback),
+      eargerly_deletion_objects_(),
+      num_bytes_in_use_(0) {}
+
+std::pair<const LocalObject *, flatbuf::PlasmaError> ObjectLifecycleManager::CreateObject(
+    const ray::ObjectInfo &object_info, plasma::flatbuf::ObjectSource source,
+    bool fallback_allocator) {
+  RAY_LOG(DEBUG) << "attempting to create object " << object_info.object_id << " size "
+                 << object_info.data_size;
+  if (object_store_->GetObject(object_info.object_id) != nullptr) {
+    return {nullptr, PlasmaError::ObjectExists};
+  }
+  auto entry = CreateObjectInternal(object_info, source, fallback_allocator);
+
+  if (entry == nullptr) {
+    return {nullptr, PlasmaError::OutOfMemory};
+  }
+  eviction_policy_->ObjectCreated(object_info.object_id, /*is_create=*/true);
+  return {entry, PlasmaError::OK};
+}
+
+const LocalObject *ObjectLifecycleManager::GetObject(const ObjectID &object_id) const {
+  return object_store_->GetObject(object_id);
+}
+
+const LocalObject *ObjectLifecycleManager::SealObject(const ObjectID &object_id) {
+  // TODO(scv119): should we check delete object from eargerly_deletion_objects_?
+  return object_store_->SealObject(object_id);
+}
+
+bool ObjectLifecycleManager::AbortObject(const ObjectID &object_id) {
+  auto entry = object_store_->GetObject(object_id);
+  if (entry == nullptr) {
+    RAY_LOG(ERROR) << "To abort an object it must be in the object table.";
+    return false;
+  }
+  if (entry->state == ObjectState::PLASMA_SEALED) {
+    RAY_LOG(ERROR) << "To abort an object it must not have been sealed.";
+    return false;
+  }
+  if (entry->ref_count > 0) {
+    // A client was using this object.
+    num_bytes_in_use_ -= entry->GetObjectSize();
+    RAY_LOG(DEBUG) << "Erasing object " << object_id << " with nonzero ref count"
+                   << object_id << ", num bytes in use is now " << num_bytes_in_use_;
+  }
+
+  DeleteObjectInternal(object_id);
+  return true;
+}
+
+PlasmaError ObjectLifecycleManager::DeleteObject(const ObjectID &object_id) {
+  auto entry = object_store_->GetObject(object_id);
+  if (entry == nullptr) {
+    return PlasmaError::ObjectNonexistent;
+  }
+
+  // TODO(scv119): should we delete unsealed with ref_count 0?
```

Comment:
The below comment suggests that this shouldn't happen?

Reply:
I think we probably should fix this to make object_lifecycle_manager has coherent APIs.